lua/luasnip/nodes/dynamicNode.lua

```diff
@@ -12,20 +11,90 @@ local log = require("luasnip.util.log").new("dynamicNode")
 local describe = require("luasnip.util.log").describe
 local session = require("luasnip.session")
 
-local function D(pos, fn, args, opts)
-	opts = opts or {}
-
-	return DynamicNode:new({
+---@class LuaSnip.SnippetNodeForDynNode: LuaSnip.SnippetNode
+---@field old_state? any
+---@field dynamicNode? LuaSnip.DynamicNode
+
+---@alias LuaSnip.DynamicNode.Fn fun(args: (string[])[], parent: LuaSnip.Snippet | LuaSnip.SnippetNode, old_state?: table, ...: table): LuaSnip.SnippetNode
+
+---@class LuaSnip.DynamicNode: LuaSnip.Node
+---@field fn LuaSnip.DynamicNode.Fn
+---@field user_args any[] Additional args that will be passed to `fn`
+---@field args LuaSnip.NodeRef[]
+---@field snip? LuaSnip.SnippetNodeForDynNode
+---@field static_snip? LuaSnip.SnippetNodeForDynNode
+---@field last_static_args? (string[])[]
+---@field snippetstring_args boolean
+local DynamicNode = Node:new()
+
+---@class LuaSnip.Opts.DynamicNode: LuaSnip.Opts.FunctionNode
+---@field snippetstring_args? boolean (FIXME: not documented?)
```

Comment:
Ah, `snippetstring_args` is a recent addition in `self-dependent-dNode`: it's a boolean which, when set, makes the dynamicNode pass `snippetStrings` instead of `string[]` as the first argument (snippetStrings are text copied from the buffer, with snippets preserved, so you can do stuff like `:gsub` on a nested snippet)

A bit of a problem with that the signature of `LuaSnip.DynamicNode.Fn` changes based on that boolean.. Can we properly represent that or can we only do something like `args:(string[]|LuaSnip.SnippetString)`?

Reply:
yeah it should be possible using `@overload` I think